### PR TITLE
Implementing all messages handling in mutlitest App

### LINKED ISSUE #388: Use builder pattern for App init
Extending #387 

With the generic modules, we no longer have generic initialisation functions (like setting the balance).

I propose the following builder pattern.

`Router::new()` // this will set with BankKeeper, WasmKeeper, and the rest unimplemented
`.with_staking(stakingKeeper)` // so we can extend these later
`.build_app()` //  this will create an app with MockApi and MockStorage

The main question I have is how to call the setup/init methods on custom keepers that go beyond the default module types (eg. init_balance). Here are some examples:

```rust
    // Add a new contract. Must be done on the base object, when no contracts running
    fn store_code(&mut self, code: Box<dyn Contract<C>>) -> usize;

    // Admin interface
    fn init_balance(
        &self,
        storage: &mut dyn Storage,
        account: &Addr,
        amount: Vec<Coin>,
    ) -> Result<(), String>;
```

We can either just set these on the router when building (and then pass the storage into the build_app constructor:

```rust
let mut router = Router::new();
let mut storage = MockStorage::new();
router.bank.init_balance(&mut storage, &owner, coins(1000, "etc"));
let mut app = router.build_app(storage);
```

Or we can expose these with some closure in the app.

```rust
let mut app = Router::new().build_app();
app.setup(|router: &Router, storage: &mut dyn Storage| {
  router.bank.init_balance(storage, &owner, coins(1000, "etc"));
  router.bank.init_balance(storage, &trader, coins(10, "btc"));
})
```

In either case, we must make the different keepers public on the router interface. I lean towards the second one.

### PR DESCRIPTION
**Pull request by hashedone**
Closes #388

### CODE REVIEW

File: packages/multi-test/src/wasm.rs [hunk 1]
```diff
@@ -861,8 +868,12 @@ mod test {
...
     }
 
-    fn mock_router() -> Router<Empty> {
-        Router::new(BankKeeper {})
+    fn mock_router() -> Router<Empty, Empty> {
+        Router {
+            wasm: Box::new(WasmKeeper::new()),
+            bank: Box::new(BankKeeper::new()),
+            custom: Box::new(MockSimpleCustomHandler::new()),
+        }
```
Comment (maurolacy) [inline packages/multi-test/src/wasm.rs:876]: Does this calls for using the builder pattern for `Router` too?

**Review by maurolacy — COMMENTED** [review]
Looks good.

Handling this into two separate MRs (builder pattern for `App` and custom message handling) would have made reviewing easy for me.

File: packages/multi-test/src/app.rs [hunk 2]
```diff
@@ -26,21 +30,26 @@ pub fn next_block(block: &mut BlockInfo) {
...
-/// C is the custom message returned init, handle, sudo (Response<C>).
-/// All contracts must return Response<C> or Response<Empty>
-pub struct App<C = Empty>
+/// ExecC is the custom message returned init, handle, sudo (Response<C>).
+/// All contracts must return Response<C> or Response<Empty>.
+///
+/// Also `ExecC` is the custom message which is handled by custom message handler.
+///
+/// `QueryC` is custom query message handled by custom message handler.
+pub struct App<ExecC = Empty, QueryC = Empty>
```
[the remark this answers is not in the document]
Comment (hashedone) [reply packages/multi-test/src/app.rs:39]: I am basically enemy of single letter names, they are semanticless. Sometimes in very obvious cases `T` makes its jobs, but besides I find it wasting more time reading, that saving writing.

File: packages/multi-test/src/test_helpers/contracts/echo.rs [hunk 3]
```diff
@@ -7,37 +7,49 @@ use cosmwasm_std::{
...
-use std::fmt;
+use std::fmt::Debug;
 
-#[derive(Debug, Clone, Serialize, Deserialize, Default)]
-pub struct Message {
+use derivative::Derivative;
+
+#[derive(Debug, Clone, Serialize, Deserialize, Derivative)]
+#[derivative(Default(bound = "", new = "true"))]
+pub struct Message<ExecC>
```
[the remark this answers is not in the document]
Comment (hashedone) [reply packages/multi-test/src/test_helpers/contracts/echo.rs:20]: It is possible for sure and I think it is even used. Empty message is still a message, good enough for testing purposes.

File: packages/multi-test/src/wasm.rs (hunk 1 above)
Comment (hashedone) [reply packages/multi-test/src/wasm.rs:876]: I don't think it makes sense, as Router is basically internal tool. Here it is the only place it is constructed besides app, and it is for very specific test, but in general router should never be constructed on its own so I don't see too much profit from moving those 5 lines to external builder (which I don't want to expose publicly).

**Comment by hashedone** [comment]
Yea, it went bigger than I expected :/ I wanted to start with all other messages (stargate and so), but I decided that I will wait for this to be reviewed as it would be too much spam.

File: packages/multi-test/src/custom_handler.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        api: &dyn Api,
+        storage: &dyn Storage,
+        block: &BlockInfo,
+        msg: QueryC,
+    ) -> AnyResult<Binary>;
+}
+
+/// Simplified version of `CustomHandler` having only arguments which are not app internals - they
+/// are just discarded. Usefull for simpler mocking.
+#[automock(type QueryResult = Binary;)]
```
Comment (ethanfrey) [inline packages/multi-test/src/custom_handler.rs:31]: I would feel happier removing this and just implementing the 2 needed methods, then mockall can be a dev-dependency only.

Comment (ethanfrey) [reply packages/multi-test/src/custom_handler.rs:31]: In reality, we will likely have much larger blockchain specific logic in the custom handlers and just applying mocks here doesn't make too much sense usually.

File: packages/multi-test/src/app.rs [hunk 5]
```diff
@@ -49,35 +58,115 @@ where
...
+    pub fn with_block(mut self, block: BlockInfo) -> Self {
+        assert!(
+            self.block.is_none(),
+            "Initial block info already overwritten"
+        );
+        self.block = Some(block);
+        self
+    }
+
+    pub fn build(self) -> App<ExecC, QueryC> {
```
Comment (ethanfrey) [inline packages/multi-test/src/app.rs:144]: Nice way to provide defaults that can be overridden

File: packages/multi-test/src/app.rs [hunk 6]
```diff
@@ -49,35 +58,115 @@ where
...
-    pub fn new(
-        api: impl Api + 'static,
-        block: BlockInfo,
-        bank: impl Bank + 'static,
-        storage: impl Storage + 'static,
-    ) -> Self {
+    /// Overwrites default wasm executor. Panic if already set.
+    #[track_caller]
+    pub fn with_wasm(mut self, wasm: impl Wasm<ExecC, QueryC> + 'static) -> Self {
+        assert!(self.wasm.is_none(), "Wasm executor already overwritten");
```
Comment (ethanfrey) [inline packages/multi-test/src/app.rs:93]: We don't allow overriding it twice? I guess no need, but seems odd to enforce.

File: packages/multi-test/src/app.rs [hunk 7]
```diff
@@ -162,28 +251,23 @@ where
     }
 }
 
-pub struct Router<C> {
-    pub wasm: Box<dyn Wasm<C>>,
-    pub bank: Box<dyn Bank>,
+pub struct Router<ExecC, QueryC> {
+    pub(crate) wasm: Box<dyn Wasm<ExecC, QueryC>>,
```
Comment (ethanfrey) [inline packages/multi-test/src/app.rs:255]: I would actually leave these public. But that can be a different PR when there is a need.

File: packages/multi-test/src/custom_handler.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+pub trait SimpleCustomHandler<ExecC = Empty, QueryC = Empty> {
+    type QueryResult;
+
+    fn execute(&self, block: &BlockInfo, sender: Addr, msg: ExecC) -> AnyResult<AppResponse>;
+    fn query(&self, block: &BlockInfo, msg: QueryC) -> AnyResult<Binary>;
+}
+
+impl<ExecC, QueryC, T: SimpleCustomHandler<ExecC, QueryC>> CustomHandler<ExecC, QueryC> for T {
```
Comment (ethanfrey) [inline packages/multi-test/src/custom_handler.rs:39]: I would just have this implementation return error (so you can test it works). For any real usage, we will have time to build a module to define the semantics of this. We will want to mock out the real blockchain behaviour, so I would tend against one-off mocks.

Comment (ethanfrey) [reply packages/multi-test/src/custom_handler.rs:39]: Or maybe a FailingCustomHandler, which returns a pre-defined error (set in new?)
And SuccessCustomHandler, which returns a fixed response (set in new?)

One or both just to help in your testing in other packages, but those don't even have to be public outside of this crate.

File: packages/multi-test/src/wasm.rs [hunk 9]
```diff
@@ -82,29 +83,33 @@ where
...
     ) -> AnyResult<AppResponse>;
 }
 
-pub struct WasmKeeper<C> {
+pub struct WasmKeeper<ExecC = Empty, QueryC = Empty> {
     /// code is in-memory lookup that stands in for wasm code
     /// this can only be edited on the WasmRouter, and just read in caches
-    codes: HashMap<usize, Box<dyn Contract<C>>>,
+    codes: HashMap<usize, Box<dyn Contract<ExecC>>>,
+    /// Just marker to make type elision fork when using it as `Wasm` trait
```
Comment (ethanfrey) [inline packages/multi-test/src/wasm.rs:96]: Actually, the contract should have QueryC in it's type.

Update: I looked at this code (also in cosmwasm-std) and it seems the query types are not enforced in DepsMut (which is where they are used). This is proper here, but maybe we can look at making cosmwasm-std more type safe

(TODO: @ethanfrey make an issue on cosmwasm-std)

**Review by ethanfrey — COMMENTED** [review]
In general, this looks good. I like the builder pattern and thanks for adding some Custom support.

I am rather strongly against using mockall here, as we will not need that in any real usage of multitest (a proper mock object should be provided by one chain). I support a minimal implementation of a custom handler (without mockall) that is just `pub(crate)` and used for the internal tests. 

Each blockchain can provide a proper implementation along with the bindings where they define CosmosMsg and QueryRequest extensions.

File: packages/multi-test/src/app.rs (hunk 6 above)
Comment (hashedone) [reply packages/multi-test/src/app.rs:93]: If you overwrite something twice in your builder then you made some mistake. Better to catch early.

File: packages/multi-test/src/app.rs (hunk 7 above)
Comment (hashedone) [reply packages/multi-test/src/app.rs:255]: Why? Any reason why user of this crate should in any point get access to internals of structure?

File: packages/multi-test/src/custom_handler.rs (hunk 4 above)
Comment (hashedone) [reply packages/multi-test/src/custom_handler.rs:31]: Basically this is the idea - specific blockchain impl just adds its own implementation. Auto-generated mock is just to make internal testing easier, but I will get rid of this in this place.

File: packages/multi-test/src/app.rs (hunk 7 above)
Comment (ethanfrey) [reply packages/multi-test/src/app.rs:255]: Not now, I have an idea, but the need will show up in a future PR.
Out of scope here. For now pub(crate) is good.

File: packages/multi-test/src/custom_handler.rs [hunk 10]
```diff
@@ -58,3 +58,72 @@ impl<ExecC, QueryC, T: SimpleCustomHandler<ExecC, QueryC>> CustomHandler<ExecC,
         self.query(block, msg)
     }
 }
+
+/// Custom handler implementation panicking on each call. Assuming, that unless specific behavior
+/// is implemented, custom messages should not be send.
+pub struct PanickingCustomHandler;
+
+impl<ExecC, QueryC> SimpleCustomHandler<ExecC, QueryC> for PanickingCustomHandler
```
Comment (ethanfrey) [inline packages/multi-test/src/custom_handler.rs:66]: Why even have the `SimpleCustomHandler` trait? Why not just implement `CustomHandler` directly?

Also, I think `CustomHandler` is the only type in this file that should be `pub` (rather than `pub(crate)`). These other types are only meaningful for us to test multi-test itself (not for blockchains that actually use custom types)

File: packages/multi-test/src/app.rs [hunk 11]
```diff
@@ -1594,6 +1582,13 @@ mod test {
                 &[],
             )
             .unwrap();
+
+            assert_eq!(
```
Comment (ethanfrey) [inline packages/multi-test/src/app.rs:1586]: This looks nice.

**Review by ethanfrey — APPROVED** [review]
Looks better.
I would still hide some more types (make them `pub(crate)`) to ensure/signal they are only for internal testing of multi-test

File: packages/multi-test/src/custom_handler.rs (hunk 10 above)
Comment (hashedone) [reply packages/multi-test/src/custom_handler.rs:66]: Yea, probably in contract we should not use in contracts tests. `SimpleCustomHandler` is here to make implementations kind of simpler as I assume that `app` is somehow accessible on the side, but it may be an overkill.
